### PR TITLE
GEODE-9439: Convert CertificateBuilder to sun/jdk APIs

### PR DESCRIPTION
**Pull request by jdeppe-pivotal**
- Makes it easier to de/serialize created certificates
- Remove the requirement for the Bouncy Castle library
- Java 9 has a very similar utility which we should be able
  to switch to once we move away from 8.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by Bill** [comment]
🥇

**Review by onichols-pivotal — APPROVED** [review]

**Comment by mhansonp** [comment]
I don't remember seeing this as a proprosal in the dev channel. Is this something that should be there? I am just asking the question... I am curious if this affects Spring or others.

**Comment by jdeppe-pivotal** [comment]
It's a change to a test helper class - I didn't think it would warrant any significant discussion.

File: geode-junit/build.gradle [hunk 1]
```diff
@@ -19,6 +19,10 @@ apply from: "${rootDir}/${scriptDir}/standard-subproject-configuration.gradle"
 
 apply from: "${project.projectDir}/../gradle/publish-java.gradle"
 
+compileJava {
+  options.compilerArgs << '-Xlint:-sunapi'
+  options.compilerArgs << '-XDenableSunApiLintControl'
```
Comment (Bill) [inline geode-junit/build.gradle:24]: So we've moved from Bouncy Castle, a high quality open source project, to an internal JDK class. These compiler options suppress what would be compile errors I suppose (or at least warnings) that we are illegally accessing JDK internal classes.

My research shows that in JDK 9 these classes in a `sun` package will move to a `javax` one. I suppose when that happens the JDK8 internal classes accessed here will have direct replacements available so all we'll have to do is change the includes in these files.

File: geode-junit/src/main/java/org/apache/geode/cache/ssl/CertificateBuilder.java [hunk 2]
```diff
@@ -79,16 +77,27 @@ public CertificateBuilder(int days, String algorithm) {
   }
 
   private static GeneralName dnsGeneralName(String name) {
-    return new GeneralName(GeneralName.dNSName, name);
+    try {
+      return new GeneralName(new DNSName(name));
+    } catch (IOException ex) {
+      throw new UncheckedIOException(ex);
+    }
```
Comment (Bill) [inline geode-junit/src/main/java/org/apache/geode/cache/ssl/CertificateBuilder.java:84]: The value of bundling this as into a runtime exception is that this method can still be used in contexts (e.g. Stream.map(f)`) where a method throwing a checked exception would not be allowed. Ok.

File: geode-junit/src/main/java/org/apache/geode/cache/ssl/CertificateBuilder.java [hunk 3]
```diff
@@ -146,58 +164,69 @@ public CertificateMaterial generate() {
...
+
+      // Update the algorithm, and resign.
+      algo = (AlgorithmId) cert.get(X509CertImpl.SIG_ALG);
+      info.set(CertificateAlgorithmId.NAME + "." + CertificateAlgorithmId.ALGORITHM, algo);
+      cert = new X509CertImpl(info);
+      cert.sign(privateKey, algorithm);
+
+      return cert;
+    } catch (Exception ex) {
+      throw new RuntimeException("Unable to create certificate", ex);
```
Comment (Bill) [inline geode-junit/src/main/java/org/apache/geode/cache/ssl/CertificateBuilder.java:229]: I'm counting on our test suite to catch any regressions in this method.

**Review by Bill — APPROVED** [review]
Nice one @jdeppe-pivotal !

**Review by ladyVader — CHANGES_REQUESTED** [review]
I think you will still need to make the suggested changes to CertificateMaterial (Optional field is not serializable).  See lines marked with =>

public class CertificateMaterial implements java.io.Serializable {
  private final X509Certificate certificate;
  private final KeyPair keyPair;
=>  private final X509Certificate issuer;
  public CertificateMaterial(X509Certificate certificate, KeyPair keyPair, X509Certificate issuer) {
    this.certificate = certificate;
    this.keyPair = keyPair;
=>    this.issuer = issuer;
  }
  public X509Certificate getCertificate() {
    return certificate;
  }
  public PublicKey getPublicKey() {
    return keyPair.getPublic();
  }
  public PrivateKey getPrivateKey() {
    return keyPair.getPrivate();
  }
  public Optional<X509Certificate> getIssuer() {
=>    return Optional.ofNullable(issuer);
  }
}

**Review by mhansonp — APPROVED** [review]

**Comment by mhansonp** [comment]
> It's a change to a test helper class - I didn't think it would warrant any significant discussion.

Sounds good. Approved..

**Comment by jdeppe-pivotal** [comment]
@ladyVader Oops - thanks for pointing that out. Just pushed that change.

**Review by ladyVader — APPROVED** [review]

**Review by dickcav — APPROVED** [review]

**Pull request merged**